In [1]:
import pandas as pd
import re
import matplotlib.pyplot as plt
import scipy
import numpy as np
import scipy


In [2]:
data = pd.read_csv("./data.csv")
ports = data["Source Port"].unique()
ports = ports[ports !=1883]
ports = ports[ports != 34366]

data


No.        Time        Source   Destination Protocol  Length   
0       1    0.000000  192.168.15.6  45.179.91.66     MQTT      68  \
1       2    0.015587  45.179.91.66  192.168.15.6     MQTT      68   
2      27    6.268177  192.168.15.6  45.179.91.66     MQTT      89   
3      29    6.283990  192.168.15.6  45.179.91.66     MQTT     103   
4      30    6.284439  45.179.91.66  192.168.15.6     MQTT      70   
..    ...         ...           ...           ...      ...     ...   
649  6100  720.040427  45.179.91.66  192.168.15.6     MQTT      68   
650  6201  732.726933  192.168.15.6  45.179.91.66     MQTT      68   
651  6202  732.744751  45.179.91.66  192.168.15.6     MQTT      68   
652  6479  780.027078  192.168.15.6  45.179.91.66     MQTT      68   
653  6480  780.044400  45.179.91.66  192.168.15.6     MQTT      68   

     Source Port Message Identifier  Destination Port   
0          34366                NaN              1883  \
1           1883                NaN             34366   
2          41067                NaN              1883   
3          41067                1,2              1883   
4           1883                NaN             41067   
..           ...                ...               ...   
649         1883                NaN             34366   
650        41067                NaN              1883   
651         1883                NaN             41067   
652        34366                NaN              1883   
653         1883                NaN             34366   

                                                  Info  
0                                         Ping Request  
1                                        Ping Response  
2                                      Connect Command  
3    Subscribe Request (id=1) [devices-1], Publish ...  
4                                          Connect Ack  
..                                                 ...  
649                                      Ping Response  
650                                       Ping Request  
651                                      Ping Response  
652                                       Ping Request  
653                                      Ping Response  

[654 rows x 10 columns]

In [3]:
devices = {}
#  : devices será um dicionario que irá ter uma chave referente ao device e o valor será um dataframe referente aos respectivos devices 
for port in ports: 
  data = data.loc[ ~(data["Message Identifier"] == "1,2")]
  array_publish = data.loc[data["Source Port"] == port].dropna(subset=['Message Identifier'])
  array_publish_ack = data.loc[data["Destination Port"] == port].dropna(subset=['Message Identifier'])
  device = {"publish":array_publish, "publish_ack" : array_publish_ack}
  devices[port] = device

In [4]:
data = {}
df_concatenado = pd.DataFrame()
for port in ports:
    RTT = pd.DataFrame(columns=["Time_publish", "Time_ack"])
    for index, row in devices[port]["publish"].iterrows():
        id = int(row["Message Identifier"])
        time_publish = row["Time"]
        RTT.loc[id, "Size_Payload"] = 246 

        RTT.loc[id, "Time_publish"] = time_publish 

    for index, row in devices[port]["publish_ack"].iterrows():
        id = int(row["Message Identifier"])
        time_publish = row["Time"]
        RTT.loc[id, "Time_ack"] = time_publish
    RTT.loc[id, "Time_ack"] = time_publish
    RTT["RTT"]  =(RTT["Time_ack"] - RTT["Time_publish"] )* 1000
    data[port] = RTT
    df_concatenado = pd.concat([data[port], df_concatenado])
df_concatenado["RTT"].std()
media = df_concatenado["RTT"].mean()
e = 0.05 * media
n = (1.96 * (df_concatenado["RTT"].std()/e))**2
n


2.3730580827711103

In [5]:

media = df_concatenado["RTT"].mean()
e = 0.05 * media
n = (1.96 * (df_concatenado["RTT"].std()/e))**2
n

2.3730580827711103

# CÁLCULO DO TAMANHO DA AMOSTRA
***
## Variáveis quantitativas e população infinita
***
## Nívéis de confiança e significância 
O **nível de confiança** ($1 - \alpha$) representa a probabilidade de acerto da estimativa. De forma complementar o **nível de significância** ($\alpha$) expressa a probabilidade de erro da estimativa.

O **nível de confiança** representa o grau de confiabilidade do resultado da estimativa estar dentro de determinado intervalo. Quando fixamos em uma pesquisa um **nível de confiança** de 95%, por exemplo, estamos assumindo que existe uma probabilidade de 95% dos resultados da pesquisa representarem bem a realidade, ou seja, estarem corretos.

O **nível de confiança** de uma estimativa pode ser obtido a partir da área sob a curva normal como ilustrado na figura abaixo.


![alt text](https://caelum-online-public.s3.amazonaws.com/1178-estatistica-parte2/01/img007.png)

Em nosso trabalho iremos utilizar um nivel de confiança de 95%, logo abaixo iremos calcular o Z.

Esta variável $Z$ representa o afastamento em desvios padrões de um valor da variável original em relação à média.

In [6]:
#variavel nível de significancia
nivel_de_significancia = 0.05

In [7]:
#Calculando nivel de confiança
nivel_de_confiança = 1 - nivel_de_significancia
nivel_de_confiança

0.95

In [8]:
#Calculando o z com scipy
z = scipy.stats.norm.ppf(0.975)
z

1.959963984540054

Com nosso z calculado podemos seguir adiante

### $$z = 1.959963984540054 $$

Abaixo segue algumas formulas que iremos utilizar para calcular o numero de amostra

## $$n = \left(z\frac{s}{e}\right)^2$$





Onde:

$z$ = variável normal padronizada

$s$ = desvio padrão amostral

$e$ = erro inferencial


### Calculando a desvio padrão $s$

In [9]:
# Calculando o desvio padrão do nosso único dispositivo
desvio_padrao_amostral = df_concatenado["RTT"].std()
desvio_padrao_amostral

1.2527895788052383

In [10]:
#Calculando o erro da nossa amostra equivalente a 5%

media = df_concatenado["RTT"].mean()
e = 0.05 * media
e

1.5939695452260132

### Calculando o tamanho da nossa amostra

# Substituindo todos os valores na formula para calcular o numero de amostra.

In [11]:
n = (z * (desvio_padrao_amostral/e))**2
n

2.3729708725740766